File: app/src/main/java/rus/tutby/mvp/feed/FeedActivity.java
package rus.tutby.mvp.feed;

import android.content.res.Configuration;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.DrawerLayout;
import android.support. v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;

import butterknife.Bind;
import butterknife.ButterKnife;
import rus.tutby.MyApplication;
import rus.tutby.R;
import rus.tutby.adapters.provideradapters.OnlinerPagerAdapter;
import rus.tutby.adapters.provideradapters.ProviderPagerAdapter;
import rus.tutby.adapters.provideradapters.TutPagerAdapter;
import rus.tutby.database.DatabaseManager;
import rus.tutby.provider.Provider;

public class FeedActivity extends AppCompatActivity {

    private static final String TAG = "TAG";

    @Bind(R.id.tab_layout) TabLayout tabLayout;
    @Bind(R.id.my_toolbar) Toolbar toolbar;
    @Bind(R.id.pager) ViewPager viewPager;
    @Bind(R.id.drawer_layout) DrawerLayout drawerLayout;

    private ProviderPagerAdapter providerPagerAdapter;
    private ActionBarDrawerToggle drawerToggle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_feed);

        ButterKnife.bind(this);

        initToolbar();
        initTabLayout();
        initViewPager();
        initNavigationDrawer();
    }

    private void initToolbar() {
        toolbar.setTitle(getActivityTitle());
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
    }

    private void initTabLayout() {
        setTabs();

        tabLayout.setOnTabSelectedListener(
                new TabLayout.ViewPagerOnTabSelectedListener(viewPager) {
                    @Override
                    public void onTabSelected(TabLayout.Tab tab) {
                        viewPager.setCurrentItem(tab.getPosition());
                    }
                });
    }

    private void initViewPager() {
        providerPagerAdapter = getProviderPagerAdapter();
        viewPager.setAdapter(providerPagerAdapter);
        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));
    }

    private void initNavigationDrawer() {
        final String[] providers = getResources().getStringArray(R.array.providers);
        final ListView drawerList;

        drawerList = (ListView) findViewById(R.id.left_drawer);
        drawerList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                switch (position) {
                    case 0:
                        MyApplication.setProvider(Provider.TUT);
                        break;
                    case 1:
                        MyApplication.setProvider(Provider.ONLINER);
                        break;
                }

                providerPagerAdapter = getProviderPagerAdapter();
                viewPager.setAdapter(providerPagerAdapter);

                toolbar.setTitle(getActivityTitle());

                setTabs();

                drawerLayout.closeDrawers();
            }
        });
        drawerList.setAdapter(new ArrayAdapter<>(this,
                android.R.layout.simple_list_item_1, providers));

        drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, R.string.drawer_open,
                R.string.drawer_close) {

            /** Called when a drawer has settled in a completely closed state. */
            public void onDrawerClosed(View view) {
                super.onDrawerClosed(view);
            }

            /** Called when a drawer has settled in a completely open state. */
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
            }
        };

        drawerLayout.setDrawerListener(drawerToggle);

    }

    private void setTabs() {
        String categories[] = getProviderCategories();
        tabLayout.removeAllTabs();
        for (String category : categories) {
            tabLayout.addTab(tabLayout.newTab().setText(category));
        }
    }

    private ProviderPagerAdapter getProviderPagerAdapter() {
        switch (MyApplication.getProvider()) {
            case TUT:
                return new TutPagerAdapter(getSupportFragmentManager(), getProviderCategories(), getProviderUrls());
            case ONLINER:
                return new OnlinerPagerAdapter(getSupportFragmentManager(), getProviderCategories(), getProviderUrls());
            default:
                return null;
        }
    }

    private String getActivityTitle() {
        switch (MyApplication.getProvider()) {
            case TUT:
                return getResources().getString(R.string.tut);
            case ONLINER:
                return getResources().getString(R.string.onliner);
            default:
                return null;
        }
    }

    private String[] getProviderCategories() {
        switch (MyApplication.getProvider()) {
            case TUT:
                return getResources().getStringArray(R.array.tut_categories);
            case ONLINER:
                return getResources().getStringArray(R.array.onliner_categories);
        }
        return null;
    }

    private String[] getProviderUrls() {
        switch (MyApplication.getProvider()) {
            case TUT:
                return getResources().getStringArray(R.array.tut_urls);
            case ONLINER:
                return getResources().getStringArray(R.array.onliner_urls);
        }
        return null;
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        drawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        drawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (drawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/rus/tutby/ui/FeedActivity.kt
package rus.tutby.ui

import android.content.res.Configuration
import android.graphics.Color
import android.graphics.Typeface
import android.support.design.widget.TabLayout
import android.support.v4.view.ViewPager
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*

import butterknife.Bind
import butterknife.ButterKnife
import com.mikepenz.materialdrawer.AccountHeader
import com.mikepenz.materialdrawer.AccountHeaderBuilder
import com.mikepenz.materialdrawer.Drawer
import com.mikepenz.materialdrawer.DrawerBuilder
import com.mikepenz.materialdrawer.model.DividerDrawerItem
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem
import com.mikepenz.materialdrawer.model.ProfileDrawerItem
import com.mikepenz.materialdrawer.model.interfaces.IProfile
import kotlinx.android.synthetic.main.activity_feed.*
import rus.tutby.App
import rus.tutby.R
import rus.tutby.ui.adapters.provideradapters.OnlinerPagerAdapter
import rus.tutby.ui.adapters.provideradapters.ProviderPagerAdapter
import rus.tutby.ui.adapters.provideradapters.TutPagerAdapter
import rus.tutby.entity.Provider
import rus.tutby.utils.showToast
import rus.tutby.utils.toBoolean
import java.util.logging.Logger

class FeedActivity : AppCompatActivity() {

    lateinit private var providerPagerAdapter: ProviderPagerAdapter
    lateinit private var drawer: Drawer

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_feed)

        initToolbar()
        initTabLayout()
        initViewPager()
        initNavigationDrawer()
    }

    private fun initToolbar() {
        toolbar.title = activityTitle
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)
    }

    private fun initTabLayout() {
        setTabs()

        tabLayout.setOnTabSelectedListener(
                object : TabLayout.ViewPagerOnTabSelectedListener(viewPager) {
                    override fun onTabSelected(tab: TabLayout.Tab) {
                        viewPager.currentItem = tab.position
                    }
                })
    }

    private fun initViewPager() {
        providerPagerAdapter = getProviderPagerAdapter()
        viewPager.adapter = providerPagerAdapter
        viewPager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tabLayout))
    }

    private fun initNavigationDrawer() {

        val accountHeader = AccountHeaderBuilder()
                .withActivity(this)
                .withHeaderBackground(R.drawable.header)
                .addProfiles(
                        ProfileDrawerItem().withName(getString(R.string.tut)).withIcon(R.drawable.tut),
                        ProfileDrawerItem().withName(getString(R.string.onliner)).withIcon(R.drawable.onliner)
                )
                .withOnAccountHeaderProfileImageListener(object : AccountHeader.OnAccountHeaderProfileImageListener {
                    override fun onProfileImageLongClick(view: View?, profile: IProfile<*>?, current: Boolean): Boolean {
                        return false
                    }

                    override fun onProfileImageClick(view: View?, profile: IProfile<*>?, current: Boolean): Boolean {
                        return changeDrawerItems(profile, current)
                    }

                })
                .withOnAccountHeaderListener { view, iProfile, b -> changeDrawerItems(iProfile, !b) }
                .build()

        drawer = DrawerBuilder()
                .withActivity(this)
                .withAccountHeader(accountHeader)
                .withToolbar(toolbar)
                .withOnDrawerItemClickListener { view, position, drawerItem ->
                    viewPager.currentItem = position - 1
                    true
                }
                .withCloseOnClick(true)
                .build()
        setDrawerItems()
    }

    private fun changeProvider() {
        when(App.getProvider()) {
            Provider.TUT -> App.setProvider(Provider.ONLINER)
            Provider.ONLINER -> App.setProvider(Provider.TUT)
        }
    }

    private fun changeDrawerItems(profile: IProfile<*>?, current: Boolean): Boolean {
        if(!current) {
            drawer.removeAllItems()
            changeProvider()
            setDrawerItems()
            providerPagerAdapter = getProviderPagerAdapter();
            viewPager.adapter = providerPagerAdapter;
            toolbar.title = activityTitle;
            setTabs();
        }
        return false
    }

    private fun setDrawerItems() {
        for(i in 0..providerCategories.size - 1) {
            drawer.addItem(
                    PrimaryDrawerItem()
                    .withName(providerCategories[i])
                    .withIdentifier((i + 1).toLong())
                    .withTextColor(Color.BLACK)
                    .withSelectedTextColorRes(R.color.colorPrimary))
        }
        drawer.addItem(DividerDrawerItem())
        drawer.addItem(
                PrimaryDrawerItem()
                .withName(getString(R.string.action_settings))
                .withIdentifier(3)
                .withTextColor(Color.BLACK)
                .withSelectedTextColorRes(R.color.colorPrimary))
    }

    private fun setTabs() {
        val categories = providerCategories
        tabLayout.removeAllTabs()
        for (i in 0..categories.size - 1) {
            tabLayout.addTab(tabLayout.newTab().setText(categories[i]))
            val tv: TextView = ((tabLayout.getChildAt(0) as LinearLayout).getChildAt(i) as LinearLayout).getChildAt(1) as TextView
            tv.typeface = Typeface.createFromAsset(assets, "Roboto-Medium.ttf")
        }
    }

    private fun getProviderPagerAdapter(): ProviderPagerAdapter {
        when (App.getProvider()) {
            Provider.TUT -> return TutPagerAdapter(supportFragmentManager, providerCategories, providerUrls)
            Provider.ONLINER -> return OnlinerPagerAdapter(supportFragmentManager, providerCategories, providerUrls)
        }
    }

    private val activityTitle: String
        get() {
            when (App.getProvider()) {
                Provider.TUT -> return resources.getString(R.string.tut)
                Provider.ONLINER -> return resources.getString(R.string.onliner)
            }
        }

    private val providerCategories: Array<String>
        get() {
            when (App.getProvider()) {
                Provider.TUT -> return resources.getStringArray(R.array.tut_categories)
                Provider.ONLINER -> return resources.getStringArray(R.array.onliner_categories)
            }
        }

    private val providerUrls: Array<String>
        get() {
            when (App.getProvider()) {
                Provider.TUT -> return resources.getStringArray(R.array.tut_urls)
                Provider.ONLINER -> return resources.getStringArray(R.array.onliner_urls)
            }
        }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return super.onOptionsItemSelected(item)
    }
}
